supply/spec/uploader_spec.rb

```diff
@@ -203,22 +203,69 @@ def find_obbs
 
-      before do
-        Supply.config = config
-        allow(Supply::Client).to receive(:make_from_config).and_return(client)
-        allow(client).to receive(:upload_apk).with(config[:apk]).and_return(1) # newly uploaded version code
-        allow(client).to receive(:begin_edit).and_return(nil)
-        allow(client).to receive(:commit_current_edit!).and_return(nil)
+    shared_examples 'run supply to upload metadata' do
+      describe '#perform_upload' do
+        subject(:release) { OpenStruct.new({ version_codes: version_codes }) }
+        let(:client) { double('client') }
+        let(:config) { { apk_paths: version_codes.map { |v_code| "some/path/app-v#{v_code}.apk" }, metadata_path: 'supply/spec/fixtures/metadata/android', track: 'track-name' } }
+
+        before do
+          Supply.config = config
+          allow(Supply::Client).to receive(:make_from_config).and_return(client)
+          version_codes.each do |version_code|
+            allow(client).to receive(:upload_apk).with("some/path/app-v#{version_code}.apk").and_return(version_code) # newly uploaded version code
+          end
+          allow(client).to receive(:upload_changelogs).and_return(nil)
+          allow(client).to receive(:tracks).with('track-name').and_return([OpenStruct.new({ releases: [ release ] })])
+          languages.each do |lang|
+            allow(client).to receive(:listing_for_language).with(lang).and_return(Supply::Listing.new(client, lang))
+            allow(client).to receive(:update_listing_for_language).with({ language: lang, full_description: "#{lang} full description", short_description: "#{lang} short description", title: "#{lang} title", video: "#{lang} video" })
+          end
+          allow(client).to receive(:begin_edit).and_return(nil)
+          allow(client).to receive(:commit_current_edit!).and_return(nil)
+        end
+
+        it 'should update track with correct version codes and optional changelog' do
+          uploader = Supply::Uploader.new
+          expect(uploader).to receive(:update_track).with(version_codes).once
+          version_codes.each do |version_code|
+            expected_notes = []
+            languages.each do |lang|
+              expected_notes << AndroidPublisher::LocalizedText.new(
+                language: lang,
+                text: "#{lang} changelog #{with_explicit_changelogs ? version_code : -1}"
+              )
+            end
+            # check if at least one of the assignments of release_notes is what we expect
+            expect(release).to receive(:release_notes=).with(same_localizedtext_as(expected_notes))
+          end
+
+          uploader.perform_upload
+        end
       end
+    end
+
+    describe '#peform_upload with metadata and explicit changelogs' do
+      it_behaves_like 'run supply to upload metadata' do
+        let(:version_codes) { [1, 2] }
+        let(:languages) { ['en-US', 'fr-FR', 'ja-JP'] }
+        let(:with_explicit_changelogs) { true }
+      end
```

Comment:
Not sure if it's just me being used to that pattern or an actual ruby/rspec codestyle recommendation, but I'm more used to see this kind of parametric tests be implemented as parameters passed to `shared_examples`, rather than shared example called with a `do…end` block redefining `let` constants for each case.

See https://railsware.com/blog/using-configurable-shared-examples-in-rspec/ for examples of such pattern with parametrized `shared_examples`.

Reply:
thanks for pointing that out. I'll update the test.